pallets/ddc-validator/src/lib.rs

```diff
@@ -0,0 +1,949 @@
+		<BalanceOf<T> as HasCompact>::Type: Clone + Eq + PartialEq + Debug + TypeInfo + Encode,
+	{
+		// Validator submits decision for an era
+		ValidationDecision(EraIndex, T::AccountId, ValidationDecision),
+		// Set era reward points
+		EraRewardPoints(EraIndex, Vec<(T::AccountId, u64)>),
+	}
+
+	#[pallet::hooks]
+	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T>
+	where
+		<T as frame_system::Config>::AccountId: AsRef<[u8]> + UncheckedFrom<T::Hash>,
+		<BalanceOf<T> as HasCompact>::Type: Clone + Eq + PartialEq + Debug + TypeInfo + Encode,
+	{
+		fn on_initialize(block_number: T::BlockNumber) -> Weight {
+			if block_number <= 1u32.into() {
+				return Weight::from_ref_time(0)
+			}
+
+			Signal::<T>::set(Some(false));
+
+			let current_ddc_era = match ddc_staking::pallet::Pallet::<T>::current_era() {
+				Some(era) => era,
+				None => {
+					log::debug!("DDC era not set.");
+					return Weight::from_ref_time(0)
+				},
+			};
+			log::debug!("Current DDC era: {:?}.", current_ddc_era);
+
+			// Produce an assignment for the next era if it's not produced yet.
+			match Self::last_managed_era() {
+				Some(last_managed_era) if current_ddc_era < last_managed_era =>
+					return Weight::from_ref_time(0),
+				_ => (),
+			}
+
+			match Self::assign(3usize, current_ddc_era + 1) {
+				Ok(_) => <LastManagedEra<T>>::put(current_ddc_era + 1),
+				Err(e) => log::debug!("DDC validation assignment error: {:?}.", e),
+			}
+
+			Weight::from_ref_time(0)
+		}
+
+		fn offchain_worker(block_number: T::BlockNumber) {
+			// Skip if not a validator.
+			if !sp_io::offchain::is_validator() {
+				return
+			}
+
+			// Skip if DDC validation is not enabled.
+			match StorageValueRef::persistent(ENABLE_DDC_VALIDATION_KEY).get::<bool>() {
+				Ok(Some(enabled)) if enabled == true => (),
+				_ => return,
+			}
+
+			let mut should_validate_because_new_era = true;
+
+			let mut validation_lock =
+				StorageLock::<storage_lock::Time>::new(LAST_VALIDATED_ERA_KEY);
+
+			// Skip if the validation is already in progress.
+			if validation_lock.try_lock().is_err() {
+				should_validate_because_new_era = false;
+			}
+
+			let last_validated_era_storage = StorageValueRef::persistent(LAST_VALIDATED_ERA_KEY);
+			let last_validated_era = match last_validated_era_storage.get::<EraIndex>() {
+				Ok(Some(last_validated_era)) => last_validated_era,
+				_ => 0, // let's consider an absent or undecodable data as we never did a validation
+			};
+
+			let current_ddc_era = match ddc_staking::pallet::Pallet::<T>::current_era() {
+				Some(era) => era,
+				None => {
+					defensive!("DDC era not set");
+					return
+				},
+			};
+
+			// Skip if the validation is already complete for the era.
+			if current_ddc_era <= last_validated_era {
+				should_validate_because_new_era = false;
+			}
+
+			// Validation start forced externally?
+			let should_validate_because_signal = Signal::<T>::get().unwrap_or(false);
```

Comment:
given other safety checks, does it actually make sense?

Reply:
It isn't a safety check. It implements a requirement formulated as "start all the validators at the same by signal".